src/workshops/gradients/gradientEditor.ts

```diff
@@ -2,54 +2,165 @@ import * as ko from "knockout";
+        gradient.colorStops().forEach((colorStop) => {
+            colorStop.color.subscribe(this.applyChanges);
+            colorStop.length.subscribe(this.applyChanges);
+        });
     }
 
-    public addColorStop(a, event: MouseEvent): void {
-        event.preventDefault();
-        event.stopPropagation();
-        const colorStopLength = Math.floor(100 * event.offsetX / (<HTMLElement>event.target).clientWidth);
+    public initializePointer(element: HTMLElement, colorStop: ColorStopViewModel): boolean {
+        const parentRect = element.parentElement.getBoundingClientRect();
+        const length = colorStop.length();
+        const position = parentRect.width * 1.0 / 100 * length - 4;
+        element.style.left= position + "px";
+        element.style.backgroundColor = colorStop.color();
+        return true;
+    } 
 
-        this.colorStops.push({ color: "#AAAAAA", length: colorStopLength });
-        this.applyChanges();
+    public async addColor(): Promise<void> {
+        const newColor = new ColorStopViewModel(<LinearGradientColorStopContract> {
+            color: "#000000",
+            length: 0
+        })
+        newColor.color.subscribe(this.applyChanges);
+        newColor.length.subscribe(this.applyChanges);
+        this.gradientViewModel().colorStops.push(newColor);
+        this.updateOnSelect();
     }
 
     private async applyChanges(): Promise<void> {
-        const abc = getLinearGradientString({ key: "", displayName: "", direction: `${this.direction()}deg`, colorStops: this.colorStops() });
+        const gradient = this.selectedGradient();
+        const colorStops: LinearGradientColorStopContract[] = [];
 
-        this.gradientPreview({
-            gradientPreview: {
-                background: { image: abc }
-            }
+        gradient.displayName = this.gradientViewModel().displayName();
+        gradient.direction = parseFloat(this.gradientViewModel().direction()) + "deg";
+        this.gradientViewModel().colorStops().forEach((colorStop) => {
+            colorStops.push(<LinearGradientColorStopContract>{
+                color: colorStop.color(),
+                length: colorStop.length()
+            });
         });
+
+        gradient.colorStops = colorStops;
+        this.updateBackground();
+    }
+
+    private async updateBackground(): Promise<void> {
+        const style = new Style("gradient-preview");
+        style.addRules([new StyleRule("backgroundImage", getLinearGradientString(this.selectedGradient()))]);
+
+        const styleSheet = new StyleSheet();
+        styleSheet.styles.push(style);
+
+        this.gradientPreview(styleSheet);
+    }
+
+    public changeColor(obIndex: ko.Observable<number>, colorValue: string): void {
+        const index = obIndex();
+        this.gradientViewModel().colorStops()[index].color(colorValue);
+        this.updateOnSelect();
+    }
+
+    public onMouseDown(obIndex: ko.Observable<number>, element: HTMLElement, event: MouseEvent): void {
```

Comment:
In general, a view model should not have references to DOM elements, for this KO (and other frameworks) have the concept of bindingHandlers (or "directive" in Angular and Vue), so I suggest to move this logic there. Just create another binding handler and name it something like "slider". In our code base there are plenty of examples of bindingHandlers.